ipatests/test_webui/ui_driver.py

```diff
@@ -174,6 +180,35 @@ def load_config(self):
         if 'type' not in c:
             c['type'] = DEFAULT_TYPE
 
+    def generate_csr(self, cn, csr_file):
+        """ generate a csr with cn and store in csr_file"""
+        key = rsa.generate_private_key(
+            public_exponent=65537,
+            key_size=2048,
+            backend=default_backend())
+
+        with open("/root/key.pem", "wb") as f:
```

Comment:
You cannot assume that you can write to ``/root``. I'd change the function to return the key and the CSR as PEM-encoded bytes.